File: modules/react-native-status/android/src/main/java/com/statusim/module/StatusModule.java
package com.statusim.module;

import android.app.Activity;
import android.view.WindowManager;
import android.os.Bundle;
import android.os.Message;
import android.os.RemoteException;
import com.facebook.react.bridge.*;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import android.util.Log;

import java.util.HashMap;
import java.util.UUID;

class StatusModule extends ReactContextBaseJavaModule implements LifecycleEventListener, ConnectorHandler {

    private static final String TAG = "StatusModule";

    private StatusConnector status = null;

    private HashMap<String, Callback> callbacks = new HashMap<>();

    StatusModule(ReactApplicationContext reactContext) {
        super(reactContext);
        reactContext.addLifecycleEventListener(this);
    }

    @Override
    public String getName() {
        return "Status";
    }

    @Override
    public void onHostResume() {  // Actvity `onResume`
        Activity currentActivity = getCurrentActivity();
        if (currentActivity == null) {
            Log.d(TAG, "On host Activity doesn't exist");
            return;
        }
        if (status == null) {
            status = new StatusConnector(currentActivity, StatusService.class);
            status.registerHandler(this);
        }

        status.bindService();

        WritableMap params = Arguments.createMap();
        Log.d(TAG, "Send module.initialized event");
        params.putString("jsonEvent", "{\"type\":\"module.initialized\"}");
        getReactApplicationContext()
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit("gethEvent", params);
    }

    @Override
    public void onHostPause() {  // Actvity `onPause`
        if (status != null) {
            status.unbindService();
        }
    }

    @Override
    public void onHostDestroy() {  // Actvity `onDestroy`
        if (status != null) {
            status.stopNode(null);
        }
    }

    @Override
    public void onConnectorConnected() {
    }

    @Override
    public void onConnectorDisconnected() {
    }

    @Override
    public boolean handleMessage(Message message) {

        Log.d(TAG, "Received message: " + message.toString());
        boolean isClaimed = true;
        Bundle bundle = message.getData();
        String callbackIdentifier = bundle.getString(StatusConnector.CALLBACK_IDENTIFIER);
        String data = bundle.getString("data");
        Callback callback = callbacks.remove(callbackIdentifier);
        switch (message.what) {
            case StatusMessages.MSG_START_NODE:
            case StatusMessages.MSG_STOP_NODE:
            case StatusMessages.MSG_LOGIN:
            case StatusMessages.MSG_CREATE_ACCOUNT:
            case StatusMessages.MSG_RECOVER_ACCOUNT:
            case StatusMessages.MSG_COMPLETE_TRANSACTION:
            case StatusMessages.MSG_JAIL_INIT:
            case StatusMessages.MSG_JAIL_PARSE:
            case StatusMessages.MSG_JAIL_CALL:
                if (callback == null) {
                    Log.d(TAG, "Could not find callback: " + callbackIdentifier);
                } else {
                    callback.invoke(data);
                }
                break;
            case StatusMessages.MSG_GETH_EVENT:
                String event = bundle.getString("event");
                WritableMap params = Arguments.createMap();
                params.putString("jsonEvent", event);
                getReactApplicationContext().getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("gethEvent", params);
                break;
            default:
                isClaimed = false;
        }

        return isClaimed;
    }

    private boolean checkAvailability() {

        Activity currentActivity = getCurrentActivity();
        if (currentActivity == null) {
            Log.d(TAG, "Activity doesn't exist");
            return false;
        }

        if (status == null) {
            Log.d(TAG, "Status connector is null");
            return false;
        }

        return true;
    }

    // Geth

    @ReactMethod
    public void startNode(Callback callback) {
        Log.d(TAG, "startNode");
        if (!checkAvailability()) {
            callback.invoke(false);
            return;
        }

        String callbackIdentifier = createIdentifier();
        callbacks.put(callbackIdentifier, callback);

        status.startNode(callbackIdentifier);
    }

    @ReactMethod
    public void login(String address, String password, Callback callback) {
        Log.d(TAG, "login");
        if (!checkAvailability()) {
            callback.invoke(false);
            return;
        }

        String callbackIdentifier = createIdentifier();
        callbacks.put(callbackIdentifier, callback);

        status.login(callbackIdentifier, address, password);
    }

    @ReactMethod
    public void createAccount(String password, Callback callback) {
        Log.d(TAG, "createAccount");
        if (!checkAvailability()) {
            callback.invoke(false);
            return;
        }

        String callbackIdentifier = createIdentifier();
        callbacks.put(callbackIdentifier, callback);

        status.createAccount(callbackIdentifier, password);
    }

    @ReactMethod
    public void recoverAccount(String passphrase, String password, Callback callback) {
        Log.d(TAG, "recoverAccount");
        if (!checkAvailability()) {
            callback.invoke(false);
            return;
        }

        String callbackIdentifier = createIdentifier();
        callbacks.put(callbackIdentifier, callback);

        status.recoverAccount(callbackIdentifier, passphrase, password);
    }

    private String createIdentifier() {
        return UUID.randomUUID().toString();
    }

    @ReactMethod
    public void completeTransaction(String hash, String password, Callback callback) {
        Log.d(TAG, "completeTransaction");
        if (!checkAvailability()) {
            callback.invoke(false);
            return;
        }

        Log.d(TAG, "Complete transaction: " + hash);
        String callbackIdentifier = createIdentifier();
        callbacks.put(callbackIdentifier, callback);

        status.completeTransaction(callbackIdentifier, hash, password);
    }


    @ReactMethod
    public void discardTransaction(String id) {
        Log.d(TAG, "discardTransaction");
        if (!checkAvailability()) {
            return;
        }

        Log.d(TAG, "Discard transaction: " + id);
        status.discardTransaction(id);
    }

    // Jail

    @ReactMethod
    public void initJail(String js, Callback callback) {
        Log.d(TAG, "initJail");
        if (!checkAvailability()) {
            callback.invoke(false);
            return;
        }

        String callbackIdentifier = createIdentifier();
        callbacks.put(callbackIdentifier, callback);

        status.initJail(callbackIdentifier, js);
    }

    @ReactMethod
    public void parseJail(String chatId, String js, Callback callback) {
        Log.d(TAG, "parseJail");
        if (!checkAvailability()) {
            callback.invoke(false);
            return;
        }

        String callbackIdentifier = createIdentifier();
        callbacks.put(callbackIdentifier, callback);

        status.parseJail(callbackIdentifier, chatId, js);
    }

    @ReactMethod
    public void callJail(String chatId, String path, String params, Callback callback) {
        Log.d(TAG, "callJail");
        if (!checkAvailability()) {
            callback.invoke(false);
            return;
        }

        String callbackIdentifier = createIdentifier();
        callbacks.put(callbackIdentifier, callback);

        status.callJail(callbackIdentifier, chatId, path, params);
    }

    @ReactMethod
    public void setAdjustResize() {
        Log.d(TAG, "setAdjustResize");
        final Activity activity = getCurrentActivity();
        if (activity == null) {
            return;
        }

        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                activity.getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE);
            }
        });
    }

    @ReactMethod
    public void setAdjustPan() {
        Log.d(TAG, "setAdjustPan");
        final Activity activity = getCurrentActivity();
        if (activity == null) {
            return;
        }

        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                activity.getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_PAN);
            }
        });
    }

    @ReactMethod
    public void setSoftInputMode(final int mode) {
        Log.d(TAG, "setSoftInputMode");
        final Activity activity = getCurrentActivity();
        if (activity == null) {
            return;
        }

        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                activity.getWindow().setSoftInputMode(mode);
            }
        });
    }
}

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.kt
package im.status.ethereum.module

import android.app.Activity
import android.os.Build
import android.util.Log
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule
import statusgo.SignalHandler
import statusgo.Statusgo
import org.json.JSONException
import android.view.WindowManager
import org.json.JSONObject

class StatusModule(private val reactContext: ReactApplicationContext, private val rootedDevice: Boolean) : ReactContextBaseJavaModule(reactContext), LifecycleEventListener, SignalHandler {

    companion object {
        private const val TAG = "StatusModule"
        private var module: StatusModule? = null

        fun getInstance(): StatusModule? {
            return module
        }
    }

    private val utils: Utils = Utils(reactContext)
    private var background: Boolean = false

    init {
        reactContext.addLifecycleEventListener(this)
    }

    override fun getName(): String {
        return "Status"
    }

    override fun onHostResume() {
        module = this
        background = false
        Statusgo.setMobileSignalHandler(this)
    }

    override fun onHostPause() {
        background = true
    }

    override fun onHostDestroy() {
        Log.d(TAG, "******************* ON HOST DESTROY *************************")
    }

    override fun handleSignal(jsonEventString: String) {
        val params = Arguments.createMap()
        params.putString("jsonEvent", jsonEventString)
        reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java).emit("gethEvent", params)
    }

    @ReactMethod
    fun closeApplication() {
        System.exit(0)
    }

    @ReactMethod
    fun connectionChange(type: String, isExpensive: Boolean) {
        Log.d(TAG, "ConnectionChange: $type, is expensive $isExpensive")
        val params = JSONObject().apply {
            put("type", type)
            put("expensive", isExpensive)
        }

        val jsonString = params.toString()
        StatusBackendClient.executeStatusGoRequest(
            endpoint = "ConnectionChangeV2",
            requestBody = jsonString,
            statusgoFunction = { Statusgo.connectionChangeV2(jsonString) }
        )
    }

    @ReactMethod
    fun appStateChange(state: String) {
        Log.d(TAG, "AppStateChange: $state")
        val params = JSONObject().apply {
            put("state", state)
        }
        val jsonString = params.toString()
        StatusBackendClient.executeStatusGoRequest(
            endpoint = "AppStateChangeV2",
            requestBody = jsonString,
            statusgoFunction = { Statusgo.appStateChangeV2(jsonString) }
        )
    }

    @ReactMethod
    fun startLocalNotifications() {
        Log.d(TAG, "startLocalNotifications")
        StatusBackendClient.executeStatusGoRequest(
            endpoint = "StartLocalNotifications",
            requestBody = "",
            statusgoFunction = { Statusgo.startLocalNotifications() }
        )
    }

    @ReactMethod
    fun getNodeConfig(callback: Callback) {
        StatusBackendClient.executeStatusGoRequestWithCallback(
            endpoint = "GetNodeConfig",
            requestBody = "",
            statusgoFunction = { Statusgo.getNodeConfig() },
            callback = callback
        )
    }

    @ReactMethod
    fun addCentralizedMetric(request: String, callback: Callback) {
        StatusBackendClient.executeStatusGoRequestWithCallback(
            endpoint = "AddCentralizedMetric",
            requestBody = request,
            statusgoFunction = { Statusgo.addCentralizedMetric(request) },
            callback
        )
    }

    @ReactMethod
    fun toggleCentralizedMetrics(request: String, callback: Callback) {
        StatusBackendClient.executeStatusGoRequestWithCallback(
            endpoint = "ToggleCentralizedMetrics",
            requestBody = request,
            statusgoFunction = { Statusgo.toggleCentralizedMetrics(request) },
            callback
        )
    }

    @ReactMethod
    fun deleteImportedKey(keyUID: String, address: String, password: String, callback: Callback) {
        val keyStoreDir = utils.getKeyStorePath(keyUID)
        val params = JSONObject().apply {
            put("address", address)
            put("password", password)
            put("keyStoreDir", keyStoreDir)
        }
        val jsonString = params.toString()
        StatusBackendClient.executeStatusGoRequestWithCallback(
            endpoint = "DeleteImportedKeyV2",
            requestBody = jsonString,
            statusgoFunction = { Statusgo.deleteImportedKeyV2(jsonString) },
            callback
        )
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun fleets(): String {
        return StatusBackendClient.executeStatusGoRequestWithResult(
            endpoint = "Fleets",
            requestBody = "",
            statusgoFunction = { Statusgo.fleets() }
        )
    }

    override fun getConstants(): Map<String, Any>? {
        return hashMapOf(
            "is24Hour" to utils.is24Hour(),
            "model" to Build.MODEL,
            "brand" to Build.BRAND,
            "buildId" to Build.ID,
            "deviceId" to Build.BOARD
        )
    }

    @ReactMethod
    fun isDeviceRooted(callback: Callback) {
        callback.invoke(rootedDevice)
    }

    @ReactMethod
    fun deactivateKeepAwake() {
        val activity = currentActivity

        activity?.runOnUiThread {
            activity.window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        }
    }
}
